### PR TITLE
Take regexps into account when setting access-control-allow-credentials

### LINKED ISSUE #43736: CORS Regex Header Allow-Credentials Bug
### Describe the bug

When using regex for CORS origins, it seems that when there is a match between the regex and the Origin, it seems that `Access-Control-Allow-Credentials` header is false, and according to the documentation it should be true. Am I missing something? This can be fixed by declaring the property, but I was just curious if it is a bug.

### Expected behavior

The header `Access-Control-Allow-Credentials` should be true when the regex matches the Origins.

### Actual behavior

If the property `quarkus.http.cors.access-control-allow-credentials` is not mentioned and if the regex matches the origin, the header's value is false.

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by gsmet**
We used to only consider exact matches which looks like an oversight.

Fixes #43736

### CODE REVIEW

**Comment by gsmet** [comment]
Failures are unrelated (Docker rate limiting in action).

**Review by michalvavrik — COMMENTED** [review]
Changes, code and expected behavior makes sense to me, but I'll leave it to @sberyozkin because I didn't change this file before IIRC.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:143]: This check is only expected to be done only if the allowed origins are not configured to `*` (wildcard), the origin is assumed to be always matching otherwise

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:157]: I'm revisiting the docs, I think the original intention was to enable this rather sensitive response header only if a precise match on the Origin is done. If we allow the regex then it can be as wide as the configured wildcard, in which case only checking against `allowsOrigin` is sufficient.

Comment (gsmet) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:157]: Hmmm, that's the reason why I did it that way. I think the regexp and * are rather different beasts and for now I was intentionally opening the gate only for the regexp (which might be used when you have several subdomains for instance).

**Comment by sberyozkin** [comment]
Thanks Guillaume @gsmet for starting looking at it, I suggested a minor optimization.
I think the intention was to avoid loose matches with the wildcard or regexes for this specific property, but since we do allow regex for the actual Origin match, it is indeed unexpected that it has no impact on setting the allowed credentials headers.

I think we should either update the origins configuration property to recommend users to avoid wildcard or regular expressions in prod and list precise Origin values, or keep insisting on the precise Origin match only... I'm leaning toward retaining the current status quo, but it can become difficult to manage when a lot of origins are involved...

How about adding a dedicated property on the allowed credentials header, we retain the current status quo by default, only if a precise Origin match is achieved, this header is returned. If users say `access-control-allow-credentials-strict-match=false`, it is a permission to Quarkus to allow it in the response even if someone configured `allowed-origins=*`.

It may be one more property too many, but it will require users double-confirm their intention that they really want to `access-control-allow-credentials`....

**Comment by sberyozkin** [comment]
Hi @gsmet FYI, https://stackoverflow.com/questions/79175348/response-to-preflight-request-doesnt-pass-access-control-check-no-access-con, it is a bit related

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java
Comment (sberyozkin) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/cors/CORSFilter.java:157]: @gsmet What I was thinking, was that if an origin like the one suggested at https://quarkus.io/guides/security-cors#support-all-origins-in-dev-mode is retained in prod, it is a regex, but it is no different to a non-regex `*`. 

Can we try to exclude a possibility of allowing credentials if `/.*/` was used for the successful origin match check. It won't prevent users from doing other very loose expressions, but at least we will minimize a chance of users opening it up unintentionally...

**Comment by sberyozkin** [comment]
@gsmet You are right that having regexs to support various subdomains with otherwise rather complete URLs in the expressions is reasonable, I'd only like to exclude expressions like `\.*\` to minimize non-intentional allowing of JavaScript accessing credentials. I'll be happy enough if we can specifically check if `\.*\` was not used as a regex source, are you OK with it ?

**Comment by gsmet** [comment]
@sberyozkin I think I addressed your concern in the second commit. I let you check all is good for you as I'm not exactly familiar with this code :).

**Comment by gsmet** [comment]
Can you clarify what you're trying to do? I would like to avoid ending up with a Frankenstein monster and something that is hard to understand?

**Comment by sberyozkin** [comment]
@gsmet Sure, I'd like to make sure that if it is a wildcard (regex or not), then `Access-Control-Allow-Credentials` is null.

Currently you have:
```
boolean allowsOrigin = wildcardOrigin;
boolean originMatches = corsConfig.origins.isPresent() &&
        (corsConfig.origins.get().contains(origin) || isOriginAllowedByRegex(allowedOriginsRegex, origin));
```

Right now, we'll have both `allowsOrigin` and `originMatches` set to true if the origin contains `/.*/` and therefore the access to credentials will be allowed.

I'd like to add only `!wildcardOrigin` to:

```
boolean allowsOrigin = wildcardOrigin;
boolean originMatches = !wildcardOrigin && corsConfig.origins.isPresent() &&
        (corsConfig.origins.get().contains(origin) || isOriginAllowedByRegex(allowedOriginsRegex, origin));
```

It will avoid allowing credential access, and also do a tiny optimization, if the origin is set to `/.*/` then `isOriginAllowedByRegex` won't run because `allowsOrigin` is already set to true with your latest update...

I've just realized only now once I started typing that this is all I wanted to do, I was not sure what exactly to suggest...

I just wanted to save you some time as opposed to keeping adding some vague comments :-), but please add this one extra boolean check and it will be good to go

**Comment by gsmet** [comment]
Ah yes, you're right, it should be fixed, thanks for catching it.

**Comment by sberyozkin** [comment]
@gsmet If you are busy with backports, I can do it, just let me know

**Comment by sberyozkin** [comment]
Hey @gsmet I did that minor test and added a test, should be ready to go... FYI, the allow credentials header is null if the Origin is wrong, but `false` it if it is not allowed but with the valid Origin...

Let me approve - and if you happy yourself with the latest update - please merge

Cheers !

**Review by sberyozkin — APPROVED** [review]

**Comment by gsmet** [comment]
Thanks!

**Pull request merged**
